# Лекция 4. Простейшие системы массового обслуживания и их характеристики

- Одноканальная СМО с неограниченной очередью
- Многоканальная СМО с неограниченной очередью
- Одноканальная СМО с отказом
- Многоканальная СМО с отказом
- Одноканальная СМО с ограничением на длину очереди
- Многоканальная СМО с ограничением на длину очереди
- Примеры решения задач
- Описание пакета моделирования СМО queueing в системе R

## Одноканальная СМО с неограниченной очередью $M/M/1/\infty$

### Открытая система

На практике довольно часто встречаются одноканальные СМО с очередью (ЭВМ, выполняющая запросы пользователя, врач, обслуживающий пациентов и т.д.). Если количество пользователей, от которых могут поступать запросы, никак не ограничено, то мы имеем дело с открытой системой.

Пусть имеется одноканальная СМО с неограниченной очередью (ни по длине очереди, ни по времени ожидания). На эту СМО поступает поток заявок с интенсивностью $\lambda$ и поток обслуживания имеет интенсивность $\mu$ ($\lambda, \mu$ сохраняют свои значения в течение всего времени функционирования системы). 

Требуется найти финальные вероятности состояний СМО, а также характеристики ее эффективности.

Состояния системы будем нумеровать по числу заявок, находящихся в СМО:
- $S_0$ - канал свободен;
- $S_1$ - канал занят, очереди нет;
- $S_2$ - канал занят, одна заявка стоит в очереди и т.д.

Число состояний ничем не ограничено. Тогда получаем классический процесс размножения, гибели при $\lambda=\lambda_k, \mu=\mu_k$, но с бесконечным числом состояний. По всем стрелкам поток заявок с интенсивностью $\lambda$ переводит систему слева направо, а справа налево - поток обслуживания с интенсивностью $\mu$.

Если $\lambda\geq\mu$, то канал с заявками не справляется, очередь растет до бесконечности, если $\lambda<\mu$, то задача разрешима.

Обозначим $y=\frac{\lambda}{\mu}$, тогда, используя формулу (25) из лекции 3, получаем:
$$P_0=\left(1+y+y^2+y^3+\ldots+y^k+\ldots\right)^{-1}$$

Выражение $1+y+y^2+y^3+\ldots+y^k+\ldots=\frac{1}{1-y}$ представляет собой геометрическую прогрессию, при $y<1$ ряд сходится, откуда получаем:
$$P_0=1-y$$

Остальные финальные вероятности находятся по формуле (24) лекции 3:
$$\forall k=1,\ldots,\infty: P_k=y^k\cdot P_0=y^k(1-y)$$

Из полученных формул, кстати, видно, что наибольшее значение имеет вероятность $P_0$ - того, что канал будет вообще свободен. Это связано с тем, что как бы ни была загружена система с очередью, если она вообще справляется с потоком заявок, самое вероятное число заявок в системе будет равно нулю.  

Найдем **среднее число заявок в СМО** $L_{\text{сист}}$. Случайная величина - число заявок в системе, имеет возможные значения $0, 1, 2, \ldots$ с вероятностями $P_0, P_1, P_2, \ldots$. Ее математическое ожидание:
$$L_{\text{сист}}=0\cdot P_0+1\cdot P_1+2\cdot P_2+\ldots = \sum_{k=1}^{\infty}k\cdot P_k=(1-y)\cdot y\cdot \sum_{k=1}^{\infty}k\cdot y^{k-1}=(1-y)\cdot y\cdot\frac{d}{d y}\sum_{k=1}^\infty y^k$$
$$L_{\text{сист}}=(1-y)\cdot y\cdot\frac{d}{d y}\left(\frac{y}{1-y}\right)=(1-y)\cdot y\cdot\frac{1}{(1-y)^2}$$
$$L_{\text{сист}}=\frac{y}{1-y}\tag{1}$$

По формуле Литтла найдем **среднее время пребывания заявки в системе**:
$$T_{\text{сист}}=\frac{L_{\text{сист}}}{\lambda}=\frac{y}{\lambda(1-y)}=\frac{1}{\mu(1-y)}\tag{2}$$

**Среднее время ожидания обслуживания**:
$$W_{\text{оч}}=T_{\text{сист}}-\frac{1}{\mu}=\frac{1}{\mu(1-y)}-\frac{1}{\mu}=\frac{y}{\mu(1-y)}=\frac{y^2}{\lambda(1-y)}\tag{3}$$

**Средняя длина очереди**:
$$L_{\text{оч}}=\lambda\cdot W_{\text{оч}}=\frac{y^2}{(1-y)}\tag{4}$$

### Замкнутая система

Если количество запросов, которые могут поступать в систему одномоментно как-то ограничено, то мы имеем дело с замкнутой системой (или системой Энгсета). 

Например, в цехе имеется $k$ прядильных станков, требующих обслуживания (в среднем через $t_1$ мин. происходит обрыв пряжи), имеется всего один работник - прядильщица, которая восстанавливает работоспособность станков. Требуется оценить эффективность функционирования данной СМО.

В замкнутых системах у источников заявок принято различать два состояния: **активное**, когда источник может сгенерировать новую заявку на обслуживание, **пассивное**, когда источник послал в СМО заявку, но еще не получил от нее ответа. В пассивном состоянии источник не может сгенерировать новую задачу. Если количество источников заявок ограничено, то интенсивность входящего потока будет зависеть от того, сколько источников находятся в активном или пассивном состояниях. В то же время исследования показали, что эта зависимость потока заявок от состояния СМО проявляется только при небольшом числе источников заявок. Если число источников достаточно велико, то практически интенсивность общего потока заявок перестает зависеть от состояния СМО (например, количество абонентов у операторов мобильной связи хотя и ограничено, но настолько велико, что это практически не отражается на входящем потоке запросов от абонентов).

Обозначим интенсивность потока отказов одного станка через $\lambda=\frac{1}{t_1}$, тогда $k$ станков дают отказы с интенсивностью $k\cdot \lambda$. Пусть также прядильщица обслуживает станки с интенсивностью $\mu$. Случайные величины: время между последовательными поломками станка, время обслуживания станка имеют экспоненциальное распределение.

Пусть $S_0$ - состояние исправности всех станков; $S_1$ - сломался один станок, очереди нет;
$S_2$ - сломалось два станка, один в очереди и т.д.

<img src="ris1.jpeg">

(на рисунке использованы обозначения $a=\lambda, m=\mu$)

В этом случае по формуле (25) лекции 3 получаем:

$$P_0=\left(1+\frac{k\lambda}{\mu}+\frac{k(k-1)\lambda^2}{\mu^2}+\frac{k(k-1)(k-2)\lambda^3}{\mu^3}+\ldots+\frac{k(k-1)\cdot\ldots\cdot(k-m+1)\lambda^{m}}{\mu^{m}}+\ldots+\frac{k!\lambda^k}{\mu^k}\right)^{-1}$$

$$P_0=\left(\sum_{m=0}^{k}\frac{k!\lambda^m}{(k-m)!\mu^m}\right)^{-1}\tag{5}$$

$$P_1=P_0\cdot \frac{k\lambda}{\mu}$$
$$P_2=P_0\cdot \frac{k(k-1)\lambda^2}{\mu^2}$$
$$\ldots$$
$$P_m=P_0\cdot \frac{k(k-1)\ldots(k-m+1)\lambda^{m}}{\mu^{m}}$$
$$\ldots$$
$$P_k=P_0\cdot \frac{k!\lambda^k}{\mu^k}$$

Вероятность $P_m$ можно также записать в виде:
$$P_m=P_0\cdot \frac{k!\lambda^{m}}{(k-m)!\mu^{m}} \tag{6}$$

Заметим, что в данном случае система имеет конечное число состояний: $S_0, S_1, \ldots, S_k$

**Среднее число заявок в системе** равно среднему числу источников заявок, находящихся в пассивном состоянии:
$$L_{\text{сист}}=N_{\text{пас}}=k-N_{\text{акт}}=k-\frac{(1-P_0)\cdot \mu}{\lambda}\tag{7}$$
$$$$

**Средняя интенсивность суммарного входного потока**, порождаемого активными источниками, равна произведению среднего числа активных источников на интенсивность потока от каждого источника:
$$\Lambda=N_{\text{акт}}\cdot \lambda = (k-L_{\text{сист}})\cdot \lambda=(1-P_0)\cdot \mu\tag{8}$$

Поскольку все заявки будут обслужены, средняя интенсивность суммарного входного потока, т.е. среднее число поданных в единицу времени заявок, равна абсолютной пропускной способности системы:
$$\lambda'=\Lambda$$

**Среднее время пребывания заявки в системе**:

В случае замкнутых систем поток заявок имеет меняющуюся в зависимости от состояния системы интенсивность потока. Поэтому в таких системах вместо интенсивности отдельного источника заявок $\lambda$ необходимо взять
среднюю интенсивность суммарного потока заявок $\Lambda$.
$$L_{\text{сист}}=\Lambda\cdot T_{\text{сист}}\Rightarrow \\ T_{\text{сист}}=\frac{L_{\text{сист}}}{\Lambda}\tag{9}$$

**Среднее число занятых каналов** равно среднему числу заявок, находящихся под обслуживанием (находим как математическое ожидание):
$$y=0\cdot P_0+1\cdot (1-P_0)=1-P_0 \tag{9'}$$

**Среднее длина очереди** найдем как разность среднего числа заявок, находящихся в системе, и среднего числа заявок, находящихся под обслуживанием:
$$L_{\text{оч}}=L_{\text{сист}}-y=k-\frac{1-P_0}{y}-(1-P_0)=k-(1-P_0)\cdot \left(1+\frac{1}{y}\right)\tag{10}$$

**Среднее время нахождения в очереди**:
$$T_{\text{сист}}=\frac{L_{\text{оч}}}{\Lambda}\tag{11}$$

## Многоканальная СМО с неограниченной очередью $M/M/n/\infty$

Аналогично одноканальной СМО решается задача и с многоканальной СМО с неограниченной очередью и $n$ каналами. Введем следующие состояния:

$S_0$ - все каналы свободны;

$S_1$ - один канал занят, очереди нет;

$S_2$ - два канала заняты, очереди нет;

$\ldots$

$S_n$ - занято $n$ каналов, очереди нет;

$S_{n+1}$ - занято $n$ каналов, одна заявка стоит в очереди;

$\ldots$

$S_{n+r}$  - занято $n$ каналов, $r$ заявок стоит в очереди;

$\ldots$

<img src="ris2.jpeg">

Полученный граф состояний соответствует схеме размножения гибели с бесконечным количеством состояний. Естественное условие существования финальных вероятностей  - это $\frac{\lambda}{n\mu}<1$.

Применяя формулы размножения гибели, найдем финальные вероятности.

Обозначим $y=\frac{\lambda}{\mu}$, получим:

из формулы (25): 
$$P_0=\left(1+\frac{y}{1!}+\frac{y^2}{2!}+\frac{y^3}{3!}+\ldots+\frac{y^n}{n!}+\frac{y^n\cdot y}{n!n}+\frac{y^n\cdot y^2}{n!n^2}+\frac{y^n\cdot y^3}{n!n^3}+\ldots\right)^{-1}$$

$$P_0=\left(\sum_{i=0}^n\frac{y^i}{i!}+\frac{y^{n+1}}{n!n}\left(1+\frac{y}{n}+\frac{y^2}{n^2}+\ldots\right)\right)^{-1}=\left(\sum_{i=0}^n\frac{y^i}{i!}+\frac{y^{n+1}}{n!n}\frac{n}{n-y}\right)^{-1}$$

$$P_0=\left(\sum_{i=0}^n\frac{y^i}{i!}+\frac{y^{n+1}}{n!(n-y)}\right)^{-1}\tag{12}$$

из формулы (24) получаем:
$$P_1=\frac{y}{1!}\cdot P_0, P_2=\frac{y^2}{2!}\cdot P_0, \ldots, P_k=\frac{y^k}{k!}\cdot P_0, \ldots, P_n=\frac{y^n}{n!}\cdot P_0\tag{13}$$
$$P_{n+1}=\frac{y^{n+1}}{n\cdot n!}\cdot P_0, \ldots, P_{n+r}=\frac{y^{n+r}}{n^r\cdot n!}\cdot P_0\tag{14}$$

Найдем основные характеристики эффективности СМО. 

**Абсолютная пропускная способность** равна интенсивности входящего потока заявок:

$$\lambda'=\lambda$$

**Среднее число заявок в очереди**:

очередь начинает образовываться, когда занятыми оказываются все $n$ каналов:
$$L_{\text{оч}}=\sum_{r=1}^\infty r\cdot P_{n+r}=\sum_{r=1}^\infty r\cdot \frac{y^{n+r}}{n^r\cdot n!}P_0$$

$$L_{\text{оч}}=\frac{y^{n+1}}{n!\cdot n}\cdot P_0\sum_{r=1}^{\infty}\frac{r\cdot y^{r-1}}{n^{r-1}}=\frac{y^{n+1}}{n!\cdot n}\cdot P_0\cdot \frac{d}{d(y/n)}\sum_{r=1}^{\infty}\frac{y^r}{n^r}=\frac{y^{n+1}}{n!\cdot n}\cdot P_0\cdot \frac{d}{d(y/n)}\left(\frac{y/n}{1-y/n}\right)$$
$$L_{\text{оч}}=\frac{y^{n+1}}{n!\cdot n}\cdot P_0\cdot \frac{1}{(1-y/n)^2}=\frac{y^{n+1}P_0}{(n-1)!(n-y)^2}\tag{15}$$

**Среднее число заявок в системе** равно сумме среднего числа заявок в очереди и среднего числа заявок под обслуживанием (оно же среднее число занятых каналов). В свою очередь, среднее число занятых каналов при условии $\frac{\lambda}{n\mu}<1$ равно отношению интенсивности входящего потока $\lambda$ к интенсивности обслуживания одного канала $\mu$: $y=\frac{\lambda}{\mu}$:
$$L_{\text{сист}}=L_{\text{оч}}+y\tag{16}$$

По формулам Литтла легко находим:
- **среднее время нахождения заявки в очереди**:
$$W_{\text{оч}}=\frac{L_{\text{оч}}}{\lambda}\tag{17}$$
- **среднее время нахождения заявки в системе**:
$$W_{\text{сист}}=\frac{L_{\text{сист}}}{\lambda}\tag{18}$$

## Одноканальная СМО с отказом $M/M/1/0$

Классическим примером одноканальной системы с отказом в обслуживании может служить сокетное соединение с одним подключением. В тот момент, когда сокет занят взаимодействием с клиентом, остальные клиенты, пришедшие на обслуживание получают отказ, т.е. образование очереди в такой системе невозможно.

Пусть $\lambda, \mu$ - интенсивности потока заявок и обслуживания соответственно. Состояние системы $S_0$ обозначает, что канал свободен, $S_1$  означает, что канал занят обслуживанием.

<img src="ris3.jpeg">

Составляя уравнения Колмогорова для стационарного режима, получаем:
$$\begin{cases} \lambda\cdot P_0=\mu\cdot P_1 \\ \mu\cdot P_1=\lambda\cdot P_0 \end{cases}$$
Таким образом, система вырождается в одно уравнение. Учитывая условие нормировки $P_0+P_1=1$, получаем:
$$P_0=\frac{\mu}{\lambda+\mu}; P_1=\frac{\lambda}{\lambda+\mu}\tag{19}$$

Полученные значения вероятностей $P_0, P_1$ по сути определяют длительность нахождения системы в состояниях $S_0, S_1$ соответственно. Вероятность нахождения системы в состоянии $S_1$ соответствует вероятности отказа: $P_{\text{отк}}=P_1$.

**Абсолютная пропускная способность** $\lambda'=(1-P_1)\lambda=\frac{\lambda\mu}{\lambda+\mu}$

## Многоканальная СМО с отказами $M/M/n/0$

Рассмотрим одну из первых "классических" задач теории массового обслуживания. Эта задача возникла из практических нужд телефонии и была решена Эрлангом.

Задача ставится так:

*Имеется $n$ каналов, на которые поступает поток заявок с интенсивностью $\lambda$. Поток обслуживания одним каналом имеет интенсивность $\mu$. Найти финальные вероятности состояний СМО, а также характеристики эффективности:* 
- абсолютную пропускную способность (среднее число заявок, обслуживаемых в единицу времени);
- вероятность отказа;
- среднее число занятых каналов.

Введем обозначения:

$S_0$ - в СМО нет ни одной заявки;

$S_1$ - в СМО находится одна заявка (один канал занят, остальные свободны).

$\ldots$

$S_k$ - в СМО находится $k$ заявок ($k$ каналов заняты, остальные свободны).

$\ldots$

$S_n$ - в СМО находится $n$ заявок (все $n$ каналов заняты).

<img src="ris4.jpeg">

Как видно из рисунка, интенсивность потоков обслуживания увеличивается с каждым занятым каналом. Используя формулы финальных вероятностей для схемы размножения гибели, получаем:
$$P_0=\left(1+\frac{\lambda}{\mu}+\frac{\lambda^2}{2!\mu^2}+\frac{\lambda^3}{3!\mu^3}+\ldots+\frac{\lambda^k}{k!\mu^k}+\ldots+\frac{\lambda^n}{n!\mu^n}\right)^{-1}$$

$$P_0=\left(\sum_{k=0}^n\frac{\lambda^k}{k!\cdot \mu^k}\right)^{-1}=\left(\sum_{k=0}^n\frac{y^k}{k!}\right)^{-1}\tag{20}$$

$$\forall k>0: P_k=\frac{y^k}{k!}\cdot P_0\tag{21}$$

Формулы (20), (21) и называются формулами Эрланга. По финальным вероятностям можно вычислить характеристики эффективности СМО.

**Вероятность того, что заявка получит отказ** совпадает с вероятностью, что все $n$ каналов будут заняты:
$$P_{\text{отк}}=P_n=\frac{y^n}{n!}\cdot P_0\tag{22}$$

**Вероятность обслуживания заявки** $$P_{\text{обсл}}=1-P_{\text{отк}}=1-\frac{y^n}{n!}\cdot P_0\tag{23}$$

**Абсолютная пропускная способность**:
$$\lambda'=\lambda\cdot \left(1-\frac{y^n}{n!}\cdot P_0\right)\tag{24}$$
Так как каждый занятый канал в единицу времени обслуживает в среднем $\mu$ заявок, **среднее число занятых каналов** равно:
$$k=\frac{\lambda'}{\mu}=y\cdot \left(1-\frac{y^n}{n!}\cdot P_0\right)\tag{25}$$

В качестве дополнительных показателей эффективности можно рассмотреть:
- **среднее число свободных каналов**:
$$k_{\text{св}}=n-k$$
- **коэффициент загрузки каналов** - отношение среднего числа занятых каналов к общему числу имеющихся каналов:
$$k_{\text{загр}}=\frac{k}{n}$$
- **коэффициент простоя каналов**:
$$k_{\text{пр}}=1-k_{\text{загр}}$$

**Пример 1.** В вычислительный центр коллективного пользования с тремя ЭВМ поступают заказы на вычислительные работы. Если работают все три ЭВМ, то вновь поступающий заказ не принимается, и предприятие вынуждено обратиться в другой вычислительный центр. Среднее время работы с одним заказом составляет 3 часа. Интенсивность потока заявок 0.25 заявок в час. Найти предельные вероятности и показатели эффективности работы вычислительного центра.

**Решение.**

По условию задачи $n=3, \lambda=0.25, \mu=\frac{1}{3}, y=\frac{\lambda}{\mu}=\frac{3}{4}$.  Найдем предельные вероятности:

$$P_0=\left(1+\frac{3}{4}+\frac{1}{2!}\cdot \left(\frac{3}{4}\right)^2+\frac{1}{3!}\cdot \left(\frac{3}{4}\right)^3\right)^{-1}$$


In [3]:
P0:(1+3/4+1/2!*(3/4)^2+1/3!*(3/4)^3)^(-1);

                                      128
(%o2)                                 ---
                                      269

In [4]:
float(%);

(%o3)                         0.4758364312267658

В соответствии с формулой (21) рассчитываем остальные вероятности:


In [7]:
for k:1 thru 3 do print("P",k,"=",(3/4)^k/k!*P0,"=", float((3/4)^k/k!*P0));

(%o6)                                done

Таким образом, в стационарном режиме работы вычислительного центра в среднем 47.6% времени нет ни одной заявки, 35.7% времени - имеется одна заявка, 13.4% времени - заняты две ЭВМ, 3.3% времени - заняты все три ЭВМ.

Следовательно, вероятность отказа равна $$P_{\text{отк}}=P_3\approx 0.033.$$

Абсолютная пропускная способность $\lambda'=0.25\cdot(1-0.033)\approx 0.24175$ заявки.

Среднее число занятых каналов $k=\frac{0.24175}{0.33}\approx 0.725$, т.е. каждая из трех ЭВМ будет занята обслуживанием заявок в среднем $\frac{72.5}{3}\approx 24.2\%$ времени.

При оценке эффективности работы вычислительного центра нужно сопоставить доходы от выполнения заявок с потерями от простоев дорогостоящих ЭВМ и принять компромиссное решение.

## Одноканальная СМО с ограничением на длину очереди $M/M/1/m$

Ограничение на длину очереди означает, что в очереди имеется всего $m$ мест, и заявка, поступившая, когда все $m$ мест заняты, отклоняется и покидает систему необслуженной. Граф состояний для одноканальной системы с ограничением на длину очереди представлен на рисунке.

<img src="ris5.jpeg">


Как видно из рисунка, возможны следующие состояния:

$S_0$ - канал обслуживания свободен;

$S_1$ - канал обслуживания занят, очереди нет;

$S_2$ - канал обслуживания занят, в очереди одна заявка;

$\ldots$

$S_{k+1}$ - канал обслуживания занят, в очереди $k$ заявок;

$S_{m+1}$ - канал обслуживания занят, в очереди $m$ заявок.

**Вероятность отказа** в такой системе равна:

$$P_{\text{отк}}=P_{m+1}=y^{m+1}P_0\tag{26}$$

**Абсолютная пропускная способность** равна:
$$\lambda'=(1-P_{\text{отк}})\lambda=\lambda-\lambda y^{m+1}P_0\tag{27}$$

Остальные параметры находятся в зависимости от значения $y$.

Если $y=1$, то 
$$P_0=\left(1+\sum_{k=1}^{m+1}y^k\right)^{-1}=\frac{1}{m+2}\tag{28}$$
$$\forall k>0: P_k=y^{k}P_0=\frac{1}{m+2}\tag{29}$$

**Средняя длина очереди**:

$$L_{\text{оч}}=\sum_{k=1}^{m}kP_{k+1}=\frac{m(m+1)}{2}P_0=\frac{m(m+1)}{(m+2)}\tag{30}$$

**Среднее время ожидания в очереди**:
$$W_{\text{оч}}=\frac{L_{\text{оч}}}{\lambda}=\frac{m(m+1)}{\lambda(m+2)}\tag{31}$$

**Среднее число заявок, находящихся под обслуживанием**:
$$L_{\text{обс}}=0\cdot P_0+1\cdot (1-P_0)=\frac{m+1}{m+2}\tag{31'}$$

**Среднее число заявок в системе**:
$$L_{\text{сист}}=L_{\text{обс}}+L_{\text{оч}}=\frac{m+1}{m+2}+\frac{m)m+1)}{2(m+2)}=\frac{m+1}{2}\tag{32}$$
**Среднее время пребывания заявки в системе** включает в себя время обслуживания и время ожидания в очереди:
$$T_{\text{сист}}=\frac{L_{\text{сист}}}{\lambda}=\frac{m+1}{2\cdot \lambda}\tag{33}$$

Если $y\neq 1$, то

**Вероятность простоя системы**:
$$P_0=\left(\sum_{m=0}^{m+1}y^m\right)^{-1}=\left(\frac{y^{m+2}-1}{y-1}\right)^{-1}=\frac{y-1}{y^{m+2}-1}\tag{34}$$
$$\forall k>0:P_k=y^kP_0\tag{35}$$

**Среднее число заявок в очереди**
$$L_{\text{оч}}=\sum_{k=1}^{m}kP_{k+1}=\frac{y-1}{y^{m+2}-1}\cdot\sum_{k=1}^m ky^{k+1}=\frac{y^2(y-1)}{y^{m+2}-1}\frac{d}{dy}\sum_{k=1}^m y^k=\frac{y^2(y-1)}{y^{m+2}-1}\frac{d}{dy}\left(\frac{y(y^m-1)}{y-1}\right)$$

[Wolfram - сумма геометрической прогрессии](https://www.wolframalpha.com/input?i=sum%28y%5Ek%2Ck%2C1%2Cm%29)

[Wolfram - производная](https://www.wolframalpha.com/input?i=diff%28sum%28y%5Ek%2Ck%2C1%2Cm%29%2Cy%29)

$$L_{\text{оч}}=\frac{y^2(y-1)}{y^{m+2}-1}\cdot\frac{my^{m+1}-(m+1)y^m+1}{(y-1)^2}$$

$$L_{\text{оч}}=\frac{y^2}{y^{m+2}-1}\cdot\frac{my^{m+1}-(m+1)y^m+1}{(y-1)}\tag{36}$$

**Среднее время ожидания в очереди**:
$$W_{\text{оч}}=\frac{L_{\text{оч}}}{\lambda}\tag{37}$$
**Среднее число заявок в системе**:
среднее число заявок, находящихся под обслуживанием (средняя загрузка канала):
$$L_{\text{обс}}=1-P_0$$
$$L_{\text{сист}}=L_{\text{оч}}+L_{\text{обс}}\tag{38}$$
**Среднее время пребывания заявки в системе** включает в себя время обслуживания и время ожидания в очереди:
$$T_{\text{сист}}=\frac{L_{\text{сист}}}{\lambda}\tag{39}$$


## Многоканальная СМО с ограничением на длину очереди $M/M/n/m$

Пусть на вход системы массового обслуживания, имеющей $n$ каналов, поступает пуассоновский поток с интенсивностью $\lambda$. Интенсивность обслуживания каждой заявки равна  $\mu$, а максимальное число мест в очереди равно $m$.

<img src="ris6.jpeg">

$S_0$ - система свободна;

$S_1$ - занят один канал, очереди нет;

$S_2$ - занято два канала, очереди нет;

$\ldots$

$S_n$ - заняты все каналы, очереди нет;

$S_{n+1}$ - заняты все каналы, одна заявка в очереди

$\ldots$

$S_{n+m}$ - заняты все каналы, заняты все $m$ мест очереди.

Здесь, также как в одноканальных системах выделяют два случая:$\frac{y}{n}=1, \frac{y}{n}\neq 1$

**Вероятность простоя**:
$$P_0=\left(1+\frac{y}{1!}+\frac{y^2}{2!}+\ldots+\frac{y^n}{n!}+\frac{y^n\cdot y}{n!n}+\frac{y^n\cdot y^2}{n!n^2}+\frac{y^n\cdot y^3}{n!n^3}+\ldots+\frac{y^n\cdot y^m}{n!n^m}\right)^{-1}$$
$$P_0=\left(\sum_{k=0}^n \frac{y^k}{k!}+\frac{y^n}{n!}\left(\frac{y}{n}+\frac{y^2}{n^2}+\frac{y^3}{n^3}+\ldots +\frac{y^m}{n^m}\right)\right)^{-1}$$

[Wolfram - упрощения](https://www.wolframalpha.com/input?i=%28y-1%29%2F%28y%5E%28m%2B2%29-1%29*sum%28k*y%5Ek%2Ck%2C1%2Cm%29)


$$P_0=\left(\sum_{k=0}^n \frac{y^k}{k!}+\frac{y^{n+1}}{n!\cdot n}\cdot \frac{1-\left(\frac{y}{n}\right)^{m}}{1-\frac{y}{n}}\right)^{-1}\tag{40}$$
$$P_{n+1}=\frac{y^{n+1}}{n\cdot n!}\cdot P_0, \ldots, P_{n+m}=\frac{y^{n+m}}{n^m\cdot n!}\cdot P_0$$

$$P_{1}=\frac{y^{1}}{1!}\cdot P_0, \ldots, P_{n}=\frac{y^{n}}{n!}\cdot P_0\tag{41}$$

**Вероятность образования очереди**:
$$P_{n+1}+P_{n+2}+\ldots+P_{n+m}=\frac{y^n}{n!}P_0\left(\frac{y}{n}+\frac{y^2}{n^2}+\ldots+\frac{y^m}{n^m}\right)=\frac{y^{n+1}}{n!\cdot n}\cdot P_0\cdot \frac{1-\left(\frac{y}{n}\right)^m}{1-\frac{y}{n}}$$
**Вероятность отказа**:
$$P_{\text{отк}}=P_{n+m}=\frac{y^{n+m}}{n^m\cdot n!}\cdot P_0\tag{42}$$

Найдем основные характеристики эффективности СМО. 

**Абсолютная пропускная способность**:

$$\lambda'=(1-P_{\text{отк}})\lambda\tag{43}$$

**Среднее число заявок в очереди**:

очередь начинает образовываться, когда занятыми оказываются все $n$ каналов:
$$L_{\text{оч}}=\sum_{k=1}^m k\cdot P_{n+k}=\sum_{k=1}^m k\frac{y^{n+k}}{n^k\cdot n!}\cdot P_0=\frac{y^nP_0}{n!}\sum_{k=1}^m \frac{ky^k}{n^k}$$
[Wolfram - упрощение суммы](https://www.wolframalpha.com/input?i=sum%28k*y%5Ek%2Fn%5Ek%2Ck%2C1%2Cm%29)
$$L_{\text{оч}}=\frac{y^nP_0}{n!}\cdot \frac{yn^{-m}(n^{m+1}-(m+1)ny^m+my^{m+1})}{(n-y)^2}\tag{43}$$

**Среднее число занятых каналов**:
$$k=\frac{\lambda'}{\mu}=y\cdot(1-P_{\text{отк}})\tag{44}$$

**Среднее число заявок в системе** равно сумме среднего числа заявок в очереди и среднего числа заявок под обслуживанием (оно же среднее число занятых каналов):
$$L_{\text{сист}}=L_{\text{оч}}+y(1-P_{\text{отк}})\tag{45}$$

По формулам Литтла легко находим:
- **среднее время нахождения заявки в очереди**:
$$W_{\text{оч}}=\frac{L_{\text{оч}}}{\lambda}\tag{46}$$
- **среднее время нахождения заявки в системе**:
$$W_{\text{сист}}=\frac{L_{\text{сист}}}{\lambda}\tag{47}$$

## Примеры решения задач

**Задача 1.**

На вход одноканальной СМО с отказами поступает простейший поток заявок с интенсивностью $\lambda=1$. Время обслуживания  - показательное с параметром $\mu=3$. В начальный момент времени $t=0$ канал свободен. Записать и решить диференциальные уравнения Колмогорова для вероятностей состояний СМО. Найти финальные вероятности и характеристики эффективности.

**Задача 2.**

Имеется одноканальная СМО с отказами. Поток с интенсивностью $\lambda$. Время обслуживания неслучайное и в точности равно $t_{\text{обсл}}=\frac{1}{\mu}$. Найти финальные вероятности состояний системы.

**Задача 3.**

Имеется двухканальная СМО с тремя местами в очереди при $\lambda=0.6, \mu=0.2$. Определить для этой СМО финальные вероятности и характеристики эффективности.

**Задача 4.**

Прибор состоит из двух узлов. Интенсивности отказов этих узлов $\lambda=0.02$, интенсивность их восстановления $\mu=1.0$. Определить вероятности состояний прибора. 

**Задача 5.** 

Имеется двухканальная простейшая СМО с отказами. На ее вход поступает поток заявок с интенсивностью 4 заявки/ч. Среднее время обслуживания одной заявки 0.8 часов. Каждая обслуженная заявка приносит доход 4 тыс.руб. Содержание каждого канала обходится в 2 тыс.руб/ч. Выяснить, выгодно ли увеличивать число каналов о трех. 

## Описание пакета моделирования СМО queueing в системе R

In [2]:
library(queueing)

In [3]:
help.search("NewInput")

R Information

Help files with alias or concept or title matching ‘NewInput’ using
fuzzy matching:


queueing::NewInput.BnD
                        Define the inputs of a new generic Birth and
                        Death process model
  Aliases: NewInput.BnD
queueing::NewInput.CJN
                        Define the inputs of a Closed Jackson Network
  Aliases: NewInput.CJN, NewInput2.CJN, NewInput3.CJN
queueing::NewInput.MCCN
                        Define the inputs of a MultiClass Closed
                        Network
  Aliases: NewInput.MCCN
queueing::NewInput.MCMN
                        Define the inputs of a MultiClass Mixed Network
  Aliases: NewInput.MCMN
queueing::NewInput.MCON
                        Define the inputs of a MultiClass Open Network
  Aliases: NewInput.MCON
queueing::NewInput.MM1
                        Define the inputs of a new M/M/1 queueing model
  Aliases: NewInput.MM1
queueing::NewInput.MM1K
                        Define the inputs of a new M/M/1/K que

In [3]:
help(NewInput.MM1)

NewInput.MM1             package:queueing              R Documentation

_D_e_f_i_n_e _t_h_e _i_n_p_u_t_s _o_f _a _n_e_w _M/_M/_1 _q_u_e_u_e_i_n_g _m_o_d_e_l

_D_e_s_c_r_i_p_t_i_o_n:

     Define the inputs of a new M/M/1 queueing model

_U_s_a_g_e:

     NewInput.MM1(lambda=0, mu=0, n=0)
     
_A_r_g_u_m_e_n_t_s:

  lambda: arrival rate

      mu: server service rate

       n: number of customers in the system from which you want to
          obtain its probabilities. Put n=0 for a idle probability (no
          customer present in the system or system idle). With n=-1, no
          probabilities are computed

_D_e_t_a_i_l_s:

     Define the inputs of a new M/M/1 queueing model

_R_e_f_e_r_e_n_c_e_s:

     [Sixto2004] Sixto Rios Insua, Alfonso Mateos Caballero, M
     Concepcion Bielza Lozoya, Antonio Jimenez Martin (2004).
     _Investigacion Operativa. Modelos deterministicos y estocasticos_.
     Editorial Centro de Estudios Ramon Areces.

_S_e_e _A_l_s_o:

     ‘CheckInput.i_MM1’

_E_x_a_m_p_l_e_s:

     ## See example 10.3 in reference [Sixto2004] for more details.
     ## create input parameters
     i_mm1 <- NewInput.MM1(lambda=1/4, mu=1/3, n=0)

**Решение задачи 1.**

Имеем одноканальную открытую СМО с неограниченной очередью.

In [4]:
i<-NewInput.MM1(lambda=1, mu=3)

In [5]:
CheckInput(i)

In [6]:
o<-QueueingModel(i)

In [7]:
summary(o)

  lambda mu c  k  m        RO        P0        Lq        Wq X   L   W Wqq Lqq
1      1  3 1 NA NA 0.3333333 0.6666667 0.1666667 0.1666667 1 0.5 0.5 0.5 1.5

Summary Output Nomenclatures

RO (ρ): нагрузка системы

P0: вероятность простоя

Lq: среднее число заявкок в очереди

Wq: среднее время ожидания в очереди

X: пропускная способность системы

L: среднее количество заявок в системе

W: среднее время нахождения в системе

Wqq: среднее время нахождения в очереди

Lqq: среднее количество заявок в очереди

Используем формулы (1), (2), (3), (4):
$$L_{\text{сист}}=\frac{y}{1-y}$$

In [8]:
lambda <- 1; mu <- 3
y <- lambda/mu
Lsys <- y/(1-y)
print(Lsys)

[1] 0.5


$$T_{\text{сист}}=\frac{y}{\lambda (1-y)}$$

In [9]:
Tsys <- y/(lambda*(1-y))
print(Tsys)

[1] 0.5


$$W_{\text{оч}}=\frac{y^2}{\lambda(1-y)}$$

In [10]:
Woh<-y^2/(lambda*(1-y))
print(Woh)

[1] 0.1666667


$$L_{\text{оч}}=\frac{y^2}{1-y}$$

In [11]:
Loh<-y^2/((1-y))
print(Loh)

[1] 0.1666667
